modules/library/main/src/main/java/org/geotools/feature/visitor/GroupByVisitor.java

```diff
@@ -0,0 +1,454 @@
+/*
+ *    GeoTools - The Open Source Java GIS Toolkit
+ *    http://geotools.org
+ *
+ *    (C) 2016, Open Source Geospatial Foundation (OSGeo)
+ *
+ *    This library is free software; you can redistribute it and/or
+ *    modify it under the terms of the GNU Lesser General Public
+ *    License as published by the Free Software Foundation;
+ *    version 2.1 of the License.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package org.geotools.feature.visitor;
+
+import com.vividsolutions.jts.geom.Envelope;
+import com.vividsolutions.jts.geom.Geometry;
+import com.vividsolutions.jts.geom.Point;
+import org.geotools.feature.DefaultFeatureCollection;
+import org.geotools.util.Converters;
+import org.geotools.util.NullProgressListener;
+import org.geotools.util.UnsupportedImplementationException;
+import org.opengis.feature.Feature;
+import org.opengis.feature.FeatureVisitor;
+import org.opengis.feature.simple.SimpleFeature;
+import org.opengis.filter.expression.Expression;
+import org.opengis.util.ProgressListener;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Group features by one or several attributes and applies an aggregator visitor to each group.
+ */
+public class GroupByVisitor implements FeatureCalc {
+
+    /**
+     * This Enum contains all the visitors that can be used with the group by visitor.
+     */
+    public enum AggregateVisitor {
+
+        Average {
```

Comment:
Violates enum naming conventions, consider AggregateVisitor.ENUM.

These things are not really visitors, you re using them to create a visitor are you not. I do kind of see what you are doing here providing a consistent create and wrap access to associate both the feature collection visitor and the calculation result.

Is it worth factoring this out into a standalone class?